### DISCUSSION TITLE
How is UI.InteractVolume supposed to work?

### DISCUSSION THREAD

**Discussion by TopperDEL** [StereoKit/StereoKit#123] [Q&A]
I tried that one for a Button I've built - but it does not react whatever I do. I can see the Volume at the right location, but it does not trigger any state-change.

Furthermore: do you have a recommended way to create a "button-click-area"? I saw in the native part of Stereokit that you are tracking the fingers in a special way to prevent triggering when the finger dipped "too deep" into the button. I try to recreate that part for my buttons, but it is tricky.

I think I need to now how to check from where a finger entered a volume. I want to only trigger when it came from "the front". I guess it should be quite easy to get that information, but I currently can't imagine how exactly.
Do you think you can simply extend your basic functionality for a more "abstract" button-touch-area? Or do you have a better way to achieve that?

Currently my buttons do work but by just checking if the finger enters a volume leads to many false clicks be the finger crossing from the side or from behind (on its way bak) and such.

**Comment by maluoi** [comment]
I believe InteractVolume was a little buggy! I didn't use it anywhere in my code, so it wasn't getting tested either. I just fixed this and added some docs for it over in e75f60ca2415a2d61dc1e9e9b0b13494f1f729cf. I'm aiming to have another preview build up sometime today as well :)

Here are the commentdocs I just wrote for it:
### UI.InteractVolume
This watches a volume of space for pinch interaction  events! If a hand is inside the space indicated by the bounds, this function will return that hand's pinch state, as well as indicate which hand did it through the out parameter.

Note that since this only provides the hand's pinch state, it  won't give you JustActive and JustInactive notifications for  when the hand enters or leaves the volume.

**bounds**: A UI hierarchy space bounding volume.

**hand**: This will be the last hand that provides a  pinch state within this volume. That means that if both hands are pinching in this volume, it will provide the Right hand.

**returns**: This will be the pinch state of the last hand that provides a pinch state within this volume. That means that if both hands are pinching in this volume, it will provide the pinch state of the Right hand.

And here's sample code:
```CSharp
// Draw a transparent volume so the user can see this space
Vec3 volumeAt   = new Vec3(0,0.2f,-0.4f);
float volumeSize = 0.2f;
Default.MeshCube.Draw(volumeMat, Matrix.TS(volumeAt, volumeSize));

BtnState volumeState = UI.InteractVolume(new Bounds(volumeAt, Vec3.One*volumeSize), out Handed hand);
if (volumeState != BtnState.Inactive)
{
    // If it just changed interaction state, make it jump in size
    float scale = volumeState.IsChanged()
        ? 0.1f
        : 0.05f;
    Lines.AddAxis(Input.Hand(hand)[FingerId.Index, JointId.Tip].Pose, scale);
}
```

### Button Behavior

It's definitely tricky to get good button behavior working! I use a pair of tricks to get this behaving well.
1. Use a line from previous finger position to current finger position for hit testing. This prevents the user's finger from passing completely through and skipping the collision volume if it's traveling fast.
2. Use two volumes for activation. The first volume is at the very front of the button and is relatively small and thin, if the finger passes through this volume, a larger one is then used for the actual button activation. This prevents presses from behind or the side. The second volume is wider, and deeper than the first volume to ensure activation happens even if the finger is traveling at an angle. Activation is only complete when the finger travels to the end of the larger volume.

Those are the core details that make StereoKit's buttons press right! There might be some more, but those are the big ones that come to mind :)

**Reply by TopperDEL** [reply]
Ok, thanks! So InteractVolume was not what I was searching for. This is for Pinch-Gestures then.

Would it be possible to get your button-logic as a seperate building block? It would be great if I could define the size and location of a "touch zone" (the area the the finger needs to touch and that slightly gets pressed smaller until it reaches the end) and you'll handle the rest? If I could then get the current Bounds of the touch-area to render the "moveable part" of a button on my own that would be best.

I did some developments today to get a better button and it works a little bit better. But it does only react on slow clicks and does not feel like it catches all situations. Your buttons do work great, but I need to adjust the design very much. Maybe it is quite easy for you to get the - I think you call it button_behaviour - to the C#-side?

**Reply by maluoi** [reply]
I can make this accessible via C# easily, I think it should be able to deliver well enough in that case too. I had designed it as a tool to help build multiple different button types, so it could be good! I can add this to v0.3.1, and make it available in a preview build. Got a timeline you'd like this by?

I probably won't add it to the docs until I've thought about this a bit harder though. I'm just not 100% sure if it's a robust way of doing this yet. Do you have any screenshots or mockups of the buttons you're designing that you would be comfortable sharing? It would be helpful to know in what way they're different :)

**Reply by TopperDEL** [reply]
That's good to hear! For the moment I'm happy if you say this can be done quite easily from your sid. Then I can simply leave this out of my brain until your implementation is available.

My design is not quite different - I just needed to tweak some things. Basically it is a backplate, a text and a sliglty transparent "click-box". All just without rounded corners, a per-control-setable background-color and a different TextStyle.
